expression/expression.go

```diff
@@ -131,19 +131,65 @@ func getSimpleExpression(triggerExpression *TriggerExpression) (*govaluate.Evalu
 		triggerExpression.WarnValue, triggerExpression.ErrorValue, triggerExpression.TriggerType)
 }
 
-func getUserExpression(triggerExpression string) (*govaluate.EvaluableExpression, error) {
-	if expr, found := exprCache.Get(triggerExpression); found {
+func getUserExpression(triggerExpression *TriggerExpression) (*govaluate.EvaluableExpression, error) {
+	if expr, found := exprCache.Get(*triggerExpression.Expression); found {
 		return expr.(*govaluate.EvaluableExpression), nil
 	}
 
-	expr, err := govaluate.NewEvaluableExpression(triggerExpression)
+	expr, err := govaluate.NewEvaluableExpression(*triggerExpression.Expression)
+
 	if err != nil {
 		if strings.Contains(err.Error(), "Undefined function") {
 			return nil, fmt.Errorf("functions is forbidden")
 		}
 		return nil, err
 	}
-
-	exprCache.Add(triggerExpression, expr, cache.NoExpiration)
+	if err := triggerExpression.expressionExtraCheckLayer(expr.Vars()); err != nil {
+		return nil, err
+	}
+	exprCache.Add(*triggerExpression.Expression, expr, cache.NoExpiration)
 	return expr, nil
 }
+
+func (triggerExpression *TriggerExpression) expressionExtraCheckLayer(vars []string) error {
+	for _, v := range vars {
+		if _, err := triggerExpression.Get(v); err != nil {
+			return err
+		}
+	}
+	if isTernaryExpression(*triggerExpression.Expression) && !checkforColon(*triggerExpression.Expression) || !checkForEmptyStates(*triggerExpression.Expression) {
```

Comment:
I do not think that this check is complex enough to check all kinds of ternary expressions. Let's start with easy example: `?:?`. Will it pass your check? Yes. Is it correct expression? No. Check the example https://play.golang.org/p/13uxKuoi-cF

Think about nested ternary expressions especially invalid ones. Your solution will not handle it correctly. 

I propose to not to check expressions by ourselves and pass it to specialised library. I can see that call of `govaluate.NewEvaluableExpression` will check expression for correctness so probably calling it will be better that checking by ourselves.

Reply:
@litleleprikon that's why i've put the function after this call `govaluate.NewEvaluableExpression()`. So basically this function acts like another filter above the govaluate validation, which detects ignored errors of evaluate.